libraries/chain/database.cpp

```diff
@@ -3404,15 +3405,58 @@ void database::clear_expired_delegations()
       itr = delegations_by_exp.begin();
    }
 }
+#ifdef STEEM_ENABLE_SMT
+template< typename smt_balance_object_type >
+void database::adjust_smt_balance( const account_object& a, const asset& delta, bool check_balance )
+{
+   //elog( "${a} ${b} ${c}", ("a", a.name) ("b", delta.amount) ("c", delta.symbol));
+   const smt_balance_object_type* bo = 
+      find< smt_balance_object_type, by_owned_symbol >( boost::make_tuple(a.name, delta.symbol) );
+   if( bo == nullptr )
+   {
+      const auto& new_balance = create< smt_balance_object_type >( [&]( smt_balance_object_type& smt_balance )
+      {
+         smt_balance.owner = a.name;
+         smt_balance.balance = delta;
+      } );
+      bo = &new_balance;
+   }
+   else
+   {
+      modify( *bo, [&]( smt_balance_object_type& smt_balance )
+      {
+         smt_balance.balance += delta;
+      } );
+   }
 
+   if( check_balance )
+   {
+      FC_ASSERT( bo->balance.amount.value >= 0, "Insufficient SMT ${smt} funds", ("smt", delta.symbol) );
+   }
+}
```

Comment:
This method should probably automatically remove balance objects if the resulting balance is 0.